File: Graphs - DFS/399. Evaluate Division/solution.cs
public class Solution {
    public double[] CalcEquation(IList<IList<string>> equations, double[] values, IList<IList<string>> queries) {
        var graph = new Dictionary<string, Dictionary<string, double>>();
        for (int i = 0; i < equations.Count; i++) {
            string a = equations[i][0], b = equations[i][1];
            if (!graph.ContainsKey(a)) graph[a] = new Dictionary<string, double>();
            if (!graph.ContainsKey(b)) graph[b] = new Dictionary<string, double>();
            graph[a][b] = values[i];
            graph[b][a] = 1 / values[i];
        }
        var res = new double[queries.Count];
        for (int i = 0; i < queries.Count; i++) {
            var visited = new HashSet<string>();
            string a = queries[i][0], b = queries[i][1];
            if (!graph.ContainsKey(a) || !graph.ContainsKey(b)) res[i] = -1.0;
            else res[i] = DFS(graph, visited, a, b);
        }
        return res;
    }

    private double DFS(Dictionary<string, Dictionary<string, double>> graph, HashSet<string> visited, string a, string b) {
        if (a == b) return 1.0;
        visited.Add(a);
        foreach (var nei in graph[a]) {
            if (!visited.Contains(nei.Key)) {
                double product = DFS(graph, visited, nei.Key, b);
                if (product > 0) return nei.Value * product;
            }
        }
        return -1.0;
    }
}

File: Graphs - DFS/399. Evaluate Division/solution.rs
use std::collections::{HashMap, HashSet};
impl Solution {
    pub fn calc_equation(equations: Vec<Vec<String>>, values: Vec<f64>, queries: Vec<Vec<String>>) -> Vec<f64> {
        let mut graph = HashMap::new();
        for (i, equation) in equations.iter().enumerate() {
            let a = &equation[0];
            let b = &equation[1];
            graph.entry(a).or_insert(HashMap::new()).insert(b, values[i]);
            graph.entry(b).or_insert(HashMap::new()).insert(a, 1.0 / values[i]);
        }
        queries.into_iter().map(|query| {
            let a = &query[0];
            let b = &query[1];
            if !graph.contains_key(a) || !graph.contains_key(b) {
                -1.0
            } else {
                let mut visited = HashSet::new();
                Self::dfs(a, b, 1.0, &graph, &mut visited).unwrap_or(-1.0)
            }
        }).collect()
    }

    fn dfs<'a>(current: &'a String, target: &'a String, product: f64, graph: &'a HashMap<&String, HashMap<&String, f64>>, visited: &mut HashSet<&'a String>) -> Option<f64> {
        visited.insert(current);
        if current == target {
            Some(product)
        } else {
            for (next, &value) in graph[current].iter() {
                if visited.contains(next) {
                    continue;
                }
                if let Some(result) = Self::dfs(next, target, product * value, graph, visited) {
                    return Some(result);
                }
            }
            None
        }
    }
}